### ISSUE TITLE
[Performance][ROCm]: Fused bf16→fp32 router GEMM falls back to a standalone copy kernel on ROCm

### ISSUE THREAD

**Issue by mpashkovskii** [vllm-project/vllm#50267]
## Summary

The MoE router gate `GateLinear` is constructed with `out_dtype=torch.float32` (`grouped_topk` requires fp32 logits). The fused fp32-output GEMM tiers in `gate_linear.py` that fold the fp32 cast into the GEMM's BLAS epilogue are all gated on `current_platform.is_cuda()`. On ROCm the gate therefore falls through to the final tier — `F.linear` in bf16 followed by a separate `output.to(fp32)` — which emits a standalone bf16→fp32 elementwise copy kernel on every MoE layer, every step.

## Details

The relevant tier does `torch.mm(x, w.T, out_dtype=fp32)`, folding the fp32 cast into the GEMM epilogue. Its eligibility (`allow_cublas_router_gemm`) requires `allow_specialized_router_gemm`, which is CUDA + SM90/SM100 only. hipBLASLt supports the same `out_dtype` epilogue, so the CUDA-only gate is unnecessarily strict.

In a kernel trace this shows up as a standalone

```
at::native::vectorized_elementwise_kernel<..., bfloat16tofloat32_copy_kernel ...>
```

running between the router GEMM and `grouped_topk`, once per MoE layer per step.

## Proposed change

Add ROCm (bias-free, bf16 weight, fp32 out) as an alternative enabling condition for `allow_cublas_router_gemm`, in both `__init__` and `set_out_dtype` (out_dtype can be set late, so both sites must agree). CUDA behavior is unchanged — the added disjunct is `current_platform.is_rocm() and ...`, which is `False` on CUDA. The fused tier then fires on ROCm and the fp32 logits come straight out of the GEMM.

## Numerical accuracy

Verified on gfx942: `torch.mm(bf16, bf16.T, out_dtype=fp32)` matches an fp32 reference to ~6e-5.

## Scope

Router gate only. It does not touch the expert GEMMs or any other bf16→fp32 intermediate in the graph (genuine fp32 intermediates in the attention-residual path are left alone).

## Environment

- Platform: ROCm / hipBLASLt (validated on gfx942)
- File: `vllm/model_executor/layers/fused_moe/router/gate_linear.py`

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by mpashkovskii** [comment]
https://github.com/vllm-project/vllm/pull/50268 is merged. Closing the ticket. Thanks @tjtanaa
